source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,571 @@
+  connect_finished_ = true;
+  next_attempt_timer_->disableTimer();
+
+  // Close and delete any other connections.
+  auto it = connections_.begin();
+  while (it != connections_.end()) {
+    if (it->get() != &(wrapper->connection())) {
+      (*it)->removeConnectionCallbacks(*wrapper);
+      (*it)->close(ConnectionCloseType::NoFlush);
+      it = connections_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  ASSERT(connections_.size() == 1);
+  callbacks_wrappers_.clear();
+
+  // Apply post-connect state to the final socket.
+  for (const auto& cb : post_connect_state_.bytes_sent_callbacks_) {
+    connections_[0]->addBytesSentCallback(cb);
+  }
+
+  if (event == ConnectionEvent::Connected) {
+    // Apply post-connect state which is only connections which have succeeded.
+    for (auto& filter : post_connect_state_.filters_) {
+      connections_[0]->addFilter(filter);
+    }
+    for (auto& filter : post_connect_state_.write_filters_) {
+      connections_[0]->addWriteFilter(filter);
+    }
+    for (auto& filter : post_connect_state_.read_filters_) {
+      connections_[0]->addReadFilter(filter);
+    }
+    if (post_connect_state_.initialize_read_filters_.has_value() &&
+        post_connect_state_.initialize_read_filters_.value()) {
+      bool initialized = connections_[0]->initializeReadFilters();
+      ASSERT(initialized);
+    }
+    if (post_connect_state_.read_disable_count_.has_value()) {
+      for (int i = 0; i < post_connect_state_.read_disable_count_.value(); ++i) {
+        connections_[0]->readDisable(true);
+      }
+    }
+
+    if (post_connect_state_.write_buffer_.has_value()) {
+      // write_buffer_ and end_stream_ are both set together in write().
+      ASSERT(post_connect_state_.end_stream_.has_value());
+      // If a buffer limit was set, ensure that it was applied to the connection.
+      if (per_connection_state_.buffer_limits_.has_value()) {
+        ASSERT(connections_[0]->bufferLimit() == per_connection_state_.buffer_limits_.value());
+      }
+      connections_[0]->write(*post_connect_state_.write_buffer_.value(),
+                             post_connect_state_.end_stream_.value());
+    }
+  }
+
+  // Add connection callbacks after moving data from the deferred write buffer so that
+  // any high watermark notification is swallowed and not conveyed to the callbacks, since
+  // that was already delivered to the callbacks when the data was written to the buffer.
+  for (auto cb : post_connect_state_.connection_callbacks_) {
+    if (cb) {
+      connections_[0]->addConnectionCallbacks(*cb);
+    }
+  }
+}
+
+void HappyEyeballsConnectionImpl::cleanupWrapperAndConnection(ConnectionCallbacksWrapper* wrapper) {
+  for (auto it = connections_.begin(); it != connections_.end();) {
+    if (it->get() == &(wrapper->connection())) {
+      (*it)->close(ConnectionCloseType::NoFlush);
+      it = connections_.erase(it);
+    } else {
+      ++it;
+    }
+  }
```

Comment:
I think that sounds right. Anything we can do to make this clearer in the code? From reading this I was wondering if this was a bug

Maybe extract function with a bool param to decide if we should remove the callbacks or just a comment somewhere?

Reply:
Thanks for digging into this. Yeah, I agree this is not awesome so I'm happy to make it better... Just not quite sure what that should look like. 

One thing we could do would be to move the current call to `removeConnectionCallbacks()` from the opt of `onEvent()` to the top of `cleanupWrapperAndConnection()`. Then we can call `removeConnectionCallbacks` for all connections in `setUpFinalConnection()`. I think this ends up with a net readability win since the remove calls are closer to the other related logic? 

I've done this is the most recent commit. WDYT?